### PR TITLE
Bug fix: Properly error out when estimateGas fails (and stacktracing is disabled)

### PR DESCRIPTION
**Pull request by haltman-at**
Addresses #4900.

So, this one's a bit messy.  I am hoping that I have done this right.  Or as right as it can be, anyway; there's some real hackery I've put in here, and I don't think there's any way to truly do this *properly*, but I'm hoping this is as close as reasonably possible, or at the very least close enough.

The problem: If `estimateGas` failed (and no explicit gas value was set), Truffle Contract wouldn't error out, but would continue ahead with a default gas value.  This is correct behavior in the special case of stacktracing mode, but incorrect behavior generally (again, per #4900).

The basic cause of the problem is that in `execute.js`, in `getGasEstimate`, the `catch()` function simply accepted, rather than rejecting with any sort of error.  So the easy fix is to simply reject with the error.  However, this is not a satisfactory solution on its own.  The problem here is that the error one will get from `estimateGas` is *different* from the error one would get from actually sending a reverting transaction.

Now, I should be clear -- to a large extent, that's just how things are going to be now; and, as I'll get to in a bit, I ended up deleting some tests as a result.  But, ideally, we would like to include some of the same information in this new error that we had in the previous error.  Like, there should be a `reason` field if applicable, and the error message (and stack trace) should include the decoded reason... including if it's a `Panic(uint)`, not only if it's an `Error(string)`.

So, this is where things get hacky.  We take the error, and then we Frankenstein additional properties onto it (and yes, we mutate... no cloning here, how are you going to properly clone an error?).  First, we use `reason.js` to get the reason; you'll notice we pass in `{ error }`, rather than `error`, in order to hit the format that `reason.js` expects.  Then we set `error.reason = reason`.  And then -- this is where things get real hacky -- we even modify `error.message` and `error.stack` (if applicable; that's what those dollar signs are for).  That last part obviously can potentially have false positives, but I'm betting on it not really coming up too much in practice.  Relying on `message` or `stack`  is already hacky business; just distinguishing a `Panic(uint)` from an `Error(string)` that way is already not perfectly reliable, after all.

Still, the result is, I hope, OK.  By the way, I should note that this won't work very well on Ganache 2, since it won't give you the binary encoded revert message.  So there you're just stuck with the decoding that Ganache gives you.  Still, better than nothing!  Like, you will still get its decoding, nothing I do here interferes with that; although it's worth noting that `reason` won't get attached in that case; its decoding will only appear in the message and stacktrace.

This was enough to get most of the contract tests to pass.  But there were a few sticklers that I wound up simply removing (well, removing particular checks in the tests, not removing the tests wholesale).  I'm hoping this is acceptable; I'll describe the remaining ones below and why I thought they were OK to remove.

Of the five removed checks, four of them were checking that we got "intrinsic gas too low" errors in certain situations.  As best I can tell, `estimateGas` just doesn't produce this error -- after all, how would it?  (Also, uh, they were checking that we got that when a constructor reverted, rather than when the intrinsic gas was actually too low...?  What...?  I'm assuming someone can explain this but I sure don't understand.)  Since I didn't see any reasonable way to fix this, I simply removed those checks.

The other removed check was about checking that we got the receipt back in the error.  Obviously, that isn't going to be possible if you didn't actually send the transaction.  So, that had to go.

You'll notice I also altered the debugger tests... there are a number of tests of debugging failed transactions, so I added gas values to those to get the transactions to actually occur.  While I was at it, I did a bit of factoring of the gas limits used in those tests.

Additionally, while I was doing all this, I also made some minor simplifications to the code.

So... I'm hoping this is OK.  It's a bit of a mess, but it's the best I could come up with...?

### CODE REVIEW

**Comment by haltman-at** [comment]
Also it only occurred to me that instead of deleting those checks, I could duplicate those tests -- have one version without gas specified, and one with, and the one with could keep the old check in.  However when I tried that just now it didn't work for some reason.  But, I could investigate and try that again if people think it's worth having those checks!

**Comment by eggplantzzz** [comment]
It's kind of complicated but this may be related to https://github.com/trufflesuite/truffle/issues/5131

File: packages/contract/lib/execute.js
Comment (eggplantzzz) [inline packages/contract/lib/execute.js:228]: Do we not need to reject the `promiEvent` any more? I'm not sure offhand how exactly this should work. If it throws here it will just stop all execution so the `promiEvent` shouldn't matter any more?

File: packages/debugger/test/data/returnvalue.js
Comment (eggplantzzz) [inline packages/debugger/test/data/returnvalue.js:300]: Do we only deal with `gas`? Should we eventually handle `gasLimit`? [Here](https://ethereum.org/en/developers/docs/transactions/) in the params section they list `gasLimit` and not `gas`.

File: packages/debugger/test/data/returnvalue.js
Comment (eggplantzzz) [inline packages/debugger/test/data/returnvalue.js:325]: We need to do this because previously it would fail on tx execution and now it will fail on gas estimation, correct?

File: packages/contract/lib/execute.js
Comment (haltman-at) [reply packages/contract/lib/execute.js:60]: Yeah -- I agree that this way of doing things is pretty wonky.  I don't really think it makes a lot of sense that `stacktrace` is a parameter to `interfaceAdapter.estimateGas`; IMO it would make sense to have the catch here, not there.  But since it already is that way, I decided not to do extra work by changing it.  However, I can totally change it if you think that would be a good idea.

Comment (haltman-at) [reply packages/contract/lib/execute.js:60]: (See discussion below.)

File: packages/debugger/test/data/returnvalue.js
Comment (haltman-at) [reply packages/debugger/test/data/returnvalue.js:325]: Yup, exactly.  Putting in `gas` because we want the tx to happen and get reverted, rather than not to happen at all.

File: packages/contract/lib/execute.js
Comment (haltman-at) [reply packages/contract/lib/execute.js:228]: As best I can tell, simply allowing this to throw works fine.  Doing this didn't break any tests, after all.  I was not able to discern any difference between throwing and having the `promiEvent` reject.

File: packages/debugger/test/data/returnvalue.js
Comment (haltman-at) [reply packages/debugger/test/data/returnvalue.js:300]: Regarding the first, Truffle and web3 call it `gas`.  That's surprising that that documentation calls it `gasLimit`.  I've never seen it called that before.

As for the second, yes, it is.  Why this value?  I dunno, but honestly, it doesn't particularly matter.  I left the actual value unchanged from what it already was.

File: packages/contract/lib/execute.js
Comment (haltman-at) [reply packages/contract/lib/execute.js:60]: Well, no, it's not a matter of what error it would throw; it's a matter of whether stacktracing is on or not.  My point is about where that should be checked; to my mind, that should be checked here, not in `estimateGas`.

File: packages/contract/lib/execute.js
Comment (haltman-at) [reply packages/contract/lib/execute.js:228]: Oh, duh -- the reason it makes no difference is because the whole thing is wrapped in `.catch(promiEvent.reject)`!

**Comment by haltman-at** [comment]
OK, so, I went back and made a few changes to this.  Specifically:

1. Those tests I removed -- I put them back in and split them in two.  For each one, there's a no-gas-specified version with the check removed, and a version with the check still in but with gas specified.
2. However, some of those tests still weren't passing!  I had to make a change to `override.js` so that it wouldn't limit the gas when getting the reason string.  (Is that OK?  Should it be a high default instead of nothing?  IDK.)
3. Finally, inspired by `override.js`, I changed my own error-mangling to do what it does; I changed it to leave `stack` alone, to have it alter `message` in a less hacky way.

File: packages/contract/lib/execute.js
Comment (eggplantzzz) [reply packages/contract/lib/execute.js:60]: Yes that is what I was trying to say. But I'm also saying that we need to check in the `catch` to make sure it is throwing for the right reason...unless the only case in which `estimateGas` fails is when something reverts.

So if something reverts, `estimateGas` will throw, right? In the case that stacktracing is on, we can ignore the error and proceed. Will `estimateGas` ever fail for a different reason? In that case we would want to throw and not continue regardless of whether stacktracing is on. Does that seem correct?

File: packages/debugger/test/data/returnvalue.js
Comment (eggplantzzz) [reply packages/debugger/test/data/returnvalue.js:300]: Maybe we should set another constant named something like `lowTxGasLimit` or something to avoid future confusion eh?

File: packages/contract/lib/execute.js
Comment (eggplantzzz) [reply packages/contract/lib/execute.js:222]: oh haha, so it will reject anyways...cool - this is confusing

File: packages/contract/lib/execute.js
Comment (haltman-at) [reply packages/contract/lib/execute.js:60]: Oh, that's a good point, I didn't think of that.  Yeah I agree there's a potential problem there.  Perhaps that can be saved for a later PR, though, since the problem is pre-existing?  Detecting the error type may require some trickiness in general.

File: packages/debugger/test/data/returnvalue.js
Comment (haltman-at) [reply packages/debugger/test/data/returnvalue.js:300]: Yeah to my knowledge it doesn't matter... I think `gasLimit` is probably fine as a name?  Calling it `arbitraryGasLimit` doesn't sound to me like it really adds much meaningful.

File: packages/contract/lib/execute.js
Comment (haltman-at) [reply packages/contract/lib/execute.js:222]: Yeah :-/

File: packages/contract/lib/execute.js
Comment (eggplantzzz) [reply packages/contract/lib/execute.js:60]: Sounds good to me, I'll leave it up to you!

File: packages/debugger/test/data/returnvalue.js
Comment (eggplantzzz) [reply packages/debugger/test/data/returnvalue.js:300]: Yeah but I kind of feel like people (including myself) get confused enough with the gas thing as it is that it is worth making it a separate case. I guess if you really don't want to do it I'll back off.

Comment (haltman-at) [reply packages/debugger/test/data/returnvalue.js:300]: I guess I could rename it to `blockGasLimit`, if you think that would be clearer?  Of course it would then *also* be used for the transaction gas limit, but that does at least make it more explicit that we *are* reusing the block gas limit as the tx gas limit.  Does that sound good to you?

Otherwise I'd rather leave it alone; it's just test code, after all.

File: packages/contract/lib/execute.js
Comment (haltman-at) [reply packages/contract/lib/execute.js:60]: Cool, I'll leave it for a later PR, and file an issue once this is merged.

File: packages/debugger/test/data/returnvalue.js
Comment (eggplantzzz) [reply packages/debugger/test/data/returnvalue.js:300]: Why couldn't we just export two variables with different names even if they have the same value so it is clear exactly what each value is for? Doesn't it kind of seem hacky to re-use a block gas limit constant for a tx gas limit? I mean, those two terms have similar names but they are very different things.

I know it is just test code but a lot of people are going to be working on this in the future and I can foresee confusion when trying to reason about tests.

Comment (haltman-at) [reply packages/debugger/test/data/returnvalue.js:300]: > Why couldn't we just export two variables with different names even if they have the same value so it is clear exactly what each value is for?

Hm, I guess I could go with that?  Seems a bit more than is necessary, but it is at least explicit.  I'll go do that I guess.

Comment (eggplantzzz) [reply packages/debugger/test/data/returnvalue.js:300]: thanks @haltman-at!

Comment (haltman-at) [reply packages/debugger/test/data/returnvalue.js]: OK, I've gone ahead and made this change.

**Review by eggplantzzz — APPROVED** [review]

**Pull request merged**
